File: app/src/main/java/com/futsch1/medtimer/OptionsMenu.java
package com.futsch1.medtimer;

import static com.futsch1.medtimer.helpers.SafeHelperKt.safeStartActivity;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.HandlerThread;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.core.view.MenuCompat;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;

import com.futsch1.medtimer.database.FullMedicine;
import com.futsch1.medtimer.database.ReminderEvent;
import com.futsch1.medtimer.exporters.CSVEventExport;
import com.futsch1.medtimer.exporters.CSVMedicineExport;
import com.futsch1.medtimer.exporters.Export;
import com.futsch1.medtimer.exporters.PDFEventExport;
import com.futsch1.medtimer.exporters.PDFMedicineExport;
import com.futsch1.medtimer.helpers.EntityEditOptionsMenu;
import com.futsch1.medtimer.helpers.FileHelper;
import com.futsch1.medtimer.helpers.PathHelper;
import com.futsch1.medtimer.helpers.SimpleIdlingResource;
import com.futsch1.medtimer.medicine.tags.TagDataFromPreferences;
import com.futsch1.medtimer.medicine.tags.TagsFragment;
import com.futsch1.medtimer.reminders.ReminderProcessorBroadcastReceiver;

import java.io.File;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.List;

public class OptionsMenu implements EntityEditOptionsMenu {
    private final Context context;
    private final Fragment fragment;
    private final MedicineViewModel medicineViewModel;
    private final NavController navController;
    private final HandlerThread backgroundThread;
    private final ActivityResultLauncher<Intent> openFileLauncher;
    private final boolean hideFilter;
    private final SimpleIdlingResource idlingResource;
    private Menu menu;
    private BackupManager backupManager;

    public OptionsMenu(Fragment fragment, MedicineViewModel medicineViewModel, NavController navController, boolean hideFilter) {
        this.fragment = fragment;
        this.context = fragment.requireContext();
        this.medicineViewModel = medicineViewModel;
        this.navController = navController;
        this.hideFilter = hideFilter;
        this.idlingResource = new SimpleIdlingResource("OptionsMenu_" + fragment.getClass().getName());
        this.openFileLauncher = fragment.registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
            if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
                this.fileSelected(result.getData().getData());
            }
        });
        backgroundThread = new HandlerThread("Export");
        backgroundThread.start();
        idlingResource.setIdle();
    }

    public void fileSelected(Uri data) {
        new Handler(backgroundThread.getLooper()).post(() -> backupManager.fileSelected(data));
    }

    @Override
    public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
        menuInflater.inflate(R.menu.main, menu);
        MenuCompat.setGroupDividerEnabled(menu, true);
        enableOptionalIcons(menu);

        this.backupManager = new BackupManager(context, fragment.getParentFragmentManager(), menu, medicineViewModel, openFileLauncher);
        this.menu = menu;
        setupSettings();
        setupVersion();
        setupAppURL();
        setupClearEvents();
        setupExport();
        setupGenerateTestData();
        setupShowAppIntro();

        handleTagFilter();
    }

    @SuppressLint("RestrictedApi")
    public static void enableOptionalIcons(@NonNull Menu menu) {
        if (menu instanceof MenuBuilder) {
            try {
                Method m = menu.getClass().getDeclaredMethod(
                        "setOptionalIconsVisible", Boolean.TYPE);
                m.invoke(menu, true);
            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
                // Intentionally empty
            }
        }
    }

    private void setupSettings() {
        MenuItem item = menu.findItem(R.id.settings);
        item.setOnMenuItemClickListener(menuItem -> {
            try {
                navController.navigate(R.id.action_global_preferencesFragment);
                return true;
            } catch (IllegalStateException e) {
                return false;
            }
        });
    }

    private void setupVersion() {
        MenuItem item = menu.findItem(R.id.version);
        item.setTitle(context.getString(R.string.version, BuildConfig.VERSION_NAME));
    }

    private void setupAppURL() {
        MenuItem item = menu.findItem(R.id.app_url);
        item.setOnMenuItemClickListener(menuItem -> {
            Intent myIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/Futsch1/medTimer"));
            safeStartActivity(context, myIntent);
            return true;
        });
    }

    private void setupClearEvents() {
        MenuItem item = menu.findItem(R.id.clear_events);
        item.setOnMenuItemClickListener(menuItem -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(context);
            builder.setTitle(R.string.confirm);
            builder.setMessage(R.string.are_you_sure_delete_events);
            builder.setCancelable(false);
            builder.setPositiveButton(R.string.yes, (dialogInterface, i) -> {
                medicineViewModel.medicineRepository.deleteReminderEvents();
                ReminderProcessorBroadcastReceiver.requestScheduleNextNotification(context);
            });
            builder.setNegativeButton(R.string.cancel, (dialogInterface, i) -> { // Intentionally left empty
            });
            builder.show();
            return true;
        });
    }

    private void setupExport() {
        Handler handler = new Handler(backgroundThread.getLooper());

        MenuItem item = menu.findItem(R.id.export_csv);
        item.setOnMenuItemClickListener(menuItem -> {
            handler.post(() -> eventExport(true));
            return true;
        });
        item = menu.findItem(R.id.export_pdf);
        item.setOnMenuItemClickListener(menuItem -> {
            handler.post(() -> eventExport(false));
            return true;
        });
        item = menu.findItem(R.id.export_medicine_csv);
        item.setOnMenuItemClickListener(menuItem -> {
            handler.post(() -> medicineExport(true));
            return true;
        });
        item = menu.findItem(R.id.export_medicine_pdf);
        item.setOnMenuItemClickListener(menuItem -> {
            handler.post(() -> medicineExport(false));
            return true;
        });
    }

    void setupGenerateTestData() {
        MenuItem item = menu.findItem(R.id.generate_test_data);
        MenuItem itemWithEvents = menu.findItem(R.id.generate_test_data_and_events);
        if (BuildConfig.DEBUG) {
            itemWithEvents.setVisible(true);
            item.setVisible(true);
            MenuItem.OnMenuItemClickListener menuItemClickListener = menuItem -> {
                idlingResource.setBusy();
                final Handler handler = new Handler(backgroundThread.getLooper());
                handler.post(() -> {
                    medicineViewModel.medicineRepository.deleteAll();
                    GenerateTestData generateTestData = new GenerateTestData(medicineViewModel, menuItem == itemWithEvents);
                    generateTestData.generateTestMedicine();
                    ReminderProcessorBroadcastReceiver.requestScheduleNextNotification(context);
                    idlingResource.setIdle();
                });
                return true;
            };
            item.setOnMenuItemClickListener(menuItemClickListener);
            itemWithEvents.setOnMenuItemClickListener(menuItemClickListener);
        } else {
            item.setVisible(false);
            itemWithEvents.setVisible(false);
        }
    }

    private void setupShowAppIntro() {
        MenuItem item = menu.findItem(R.id.show_app_intro);
        if (BuildConfig.DEBUG) {
            item.setVisible(true);
            item.setOnMenuItemClickListener(menuItem -> {
                context.startActivity(new Intent(context, MedTimerAppIntro.class));
                return true;
            });
        } else {
            item.setVisible(false);
        }
    }

    private void handleTagFilter() {
        if (!hideFilter) {
            new Handler(backgroundThread.getLooper()).post(() -> {
                if (medicineViewModel.medicineRepository.hasTags()) {
                    try {
                        fragment.requireActivity().runOnUiThread(this::setupTagFilter);
                    } catch (IllegalStateException e) {
                        // Intentionally empty, do nothing
                    }
                } else {
                    medicineViewModel.getValidTagIds().postValue(new HashSet<>());
                }
            });
        }
    }

    private void eventExport(boolean isCSV) {
        if (medicineViewModel.tagFilterActive()) {
            Toast.makeText(context, R.string.tag_filter_active, Toast.LENGTH_LONG).show();
        }
        List<ReminderEvent> reminderEvents = medicineViewModel.filterEvents(medicineViewModel.medicineRepository.getAllReminderEventsWithoutDeletedAndAcknowledged());
        Export exporter = isCSV ? new CSVEventExport(reminderEvents, fragment.getParentFragmentManager(), context) : new PDFEventExport(reminderEvents, fragment.getParentFragmentManager(), context);
        export(exporter);
    }

    private void medicineExport(boolean isCSV) {
        if (medicineViewModel.tagFilterActive()) {
            Toast.makeText(context, R.string.tag_filter_active, Toast.LENGTH_LONG).show();
        }
        List<FullMedicine> medicines = medicineViewModel.filterMedicines(medicineViewModel.medicineRepository.getMedicines());
        Export exporter = isCSV ? new CSVMedicineExport(medicines, fragment.getParentFragmentManager(), context) : new PDFMedicineExport(medicines, fragment.getParentFragmentManager(), context);
        export(exporter);
    }

    private void setupTagFilter() {
        if (fragment.getView() == null) {
            return;
        }

        MenuItem item = menu.findItem(R.id.tag_filter);
        item.setVisible(true);
        item.setOnMenuItemClickListener(menuItem -> {
            TagDataFromPreferences tagDataFromPreferences = new TagDataFromPreferences(fragment);
            DialogFragment dialog = new TagsFragment(tagDataFromPreferences);
            dialog.show(fragment.getParentFragmentManager(), "tags");
            return true;
        });
        medicineViewModel.getValidTagIds().observe(fragment.getViewLifecycleOwner(), validTagIds -> {
            if (validTagIds.isEmpty()) {
                item.setIcon(R.drawable.tag);
            } else {
                item.setIcon(R.drawable.tag_fill);
            }
        });
    }

    private void export(Export export) {
        File csvFile = new File(context.getCacheDir(), PathHelper.getExportFilename(export));
        try {
            export.export(csvFile);
            FileHelper.shareFile(context, csvFile);
        } catch (Export.ExporterException e) {
            Toast.makeText(context, R.string.export_failed, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
        return false;
    }

    public void onDestroy() {
        backgroundThread.quit();
        idlingResource.destroy();
    }
}

File: app/src/main/java/com/futsch1/medtimer/OptionsMenu.kt
package com.futsch1.medtimer

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.view.menu.MenuBuilder
import androidx.core.net.toUri
import androidx.core.view.MenuCompat
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavController
import com.futsch1.medtimer.database.FullMedicine
import com.futsch1.medtimer.database.MedicineRepository
import com.futsch1.medtimer.database.ReminderEvent
import com.futsch1.medtimer.di.Dispatcher
import com.futsch1.medtimer.di.MedTimerDispatchers
import com.futsch1.medtimer.exporters.CSVEventExport
import com.futsch1.medtimer.exporters.CSVMedicineExport
import com.futsch1.medtimer.exporters.Export
import com.futsch1.medtimer.exporters.Export.ExporterException
import com.futsch1.medtimer.exporters.PDFEventExport
import com.futsch1.medtimer.exporters.PDFMedicineExport
import com.futsch1.medtimer.helpers.EntityEditOptionsMenu
import com.futsch1.medtimer.helpers.FileHelper
import com.futsch1.medtimer.helpers.PathHelper.getExportFilename
import com.futsch1.medtimer.helpers.SimpleIdlingResource
import com.futsch1.medtimer.helpers.safeStartActivity
import com.futsch1.medtimer.medicine.tags.TagDataFromPreferences
import com.futsch1.medtimer.medicine.tags.TagsFragment
import com.futsch1.medtimer.reminders.ReminderProcessorBroadcastReceiver.Companion.requestScheduleNextNotification
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.lang.reflect.InvocationTargetException

class OptionsMenu @AssistedInject constructor(
    @Assisted private val fragment: Fragment,
    @Assisted private val navController: NavController,
    @Assisted private val hideFilter: Boolean,
    @Assisted private val medicineViewModel: MedicineViewModel,
    val backupManagerFactory: BackupManager.Factory,
    private val pdfMedicineExportFactory: PDFMedicineExport.Factory,
    private val csvMedicineExportFactory: CSVMedicineExport.Factory,
    private val pdfEventExportFactory: PDFEventExport.Factory,
    private val csvEventExportFactory: CSVEventExport.Factory,
    private val medicineRepository: MedicineRepository,
    private val generateTestDataFactory: GenerateTestData.Factory,
    @param:Dispatcher(MedTimerDispatchers.IO) val ioDispatcher: CoroutineDispatcher,
    @param:Dispatcher(MedTimerDispatchers.Main) val mainDispatcher: CoroutineDispatcher,
    val tagsFragmentFactory: TagsFragment.Factory
) : EntityEditOptionsMenu {
    private val context: Context = fragment.requireContext()
    private val openFileLauncher: ActivityResultLauncher<Intent>
    private val openDirectoryLauncher: ActivityResultLauncher<Intent>
    private val idlingResource: SimpleIdlingResource = SimpleIdlingResource("OptionsMenu_" + fragment.javaClass.getName())
    private lateinit var menu: Menu

    private lateinit var backupManager: BackupManager

    @AssistedFactory
    fun interface Factory {
        fun create(
            fragment: Fragment,
            navController: NavController,
            hideFilter: Boolean,
            medicineViewModel: MedicineViewModel
        ): OptionsMenu
    }

    init {
        openFileLauncher =
            fragment.registerForActivityResult(StartActivityForResult()) { result ->
                if (result.resultCode == Activity.RESULT_OK) {
                    result.data?.data?.let { this.fileSelected(it) }
                }
            }
        openDirectoryLauncher =
            fragment.registerForActivityResult(StartActivityForResult()) { result ->
                val resultData = result.data ?: return@registerForActivityResult
                if (result.resultCode == Activity.RESULT_OK) {
                    this.directorySelected(resultData.data)
                }
            }
        idlingResource.setIdle()
    }

    fun fileSelected(data: Uri) {
        fragment.lifecycleScope.launch(ioDispatcher) {
            backupManager.fileSelected(data)
        }
    }

    fun directorySelected(data: Uri?) {
        backupManager.directorySelected(data)
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.main, menu)
        MenuCompat.setGroupDividerEnabled(menu, true)
        enableOptionalIcons(menu)

        this.backupManager = backupManagerFactory.create(
            fragment.requireContext(),
            fragment,
            menu,
            openFileLauncher,
            openDirectoryLauncher,
            fragment.parentFragmentManager
        )
        this.menu = menu
        setupSettings()
        setupVersion()
        setupAppURL()
        setupClearEvents()
        setupExport()
        setupGenerateTestData()
        setupShowAppIntro()

        handleTagFilter()
    }

    private fun setupSettings() {
        val item = menu.findItem(R.id.settings)
        item.setOnMenuItemClickListener { _: MenuItem? ->
            try {
                navController.navigate(R.id.action_global_preferencesFragment)
                return@setOnMenuItemClickListener true
            } catch (_: IllegalStateException) {
                return@setOnMenuItemClickListener false
            }
        }
    }

    private fun setupVersion() {
        val item = menu.findItem(R.id.version)
        item.title = context.getString(R.string.version, BuildConfig.VERSION_NAME)
    }

    private fun setupAppURL() {
        val item = menu.findItem(R.id.app_url)
        item.setOnMenuItemClickListener { _: MenuItem? ->
            val myIntent = Intent(Intent.ACTION_VIEW, "https://github.com/Futsch1/medTimer".toUri())
            safeStartActivity(context, myIntent)
            true
        }
    }

    private fun setupClearEvents() {
        val item = menu.findItem(R.id.clear_events)
        item.setOnMenuItemClickListener { _ ->
            val builder = MaterialAlertDialogBuilder(context)
            builder.setTitle(R.string.confirm)
            builder.setMessage(R.string.are_you_sure_delete_events)
            builder.setCancelable(false)
            builder.setPositiveButton(R.string.yes) { _, _ ->
                fragment.lifecycleScope.launch {
                    medicineRepository.deleteReminderEvents()
                    requestScheduleNextNotification(context)
                }
            }
            builder.setNegativeButton(R.string.cancel) { _, _ -> }
            builder.show()
            true
        }
    }

    private fun setupExport() {
        var item = menu.findItem(R.id.export_csv)
        item.setOnMenuItemClickListener { _ ->
            fragment.lifecycleScope.launch(ioDispatcher) { eventExport(true) }
            true
        }
        item = menu.findItem(R.id.export_pdf)
        item.setOnMenuItemClickListener { _ ->
            fragment.lifecycleScope.launch(ioDispatcher) { eventExport(false) }
            true
        }
        item = menu.findItem(R.id.export_medicine_csv)
        item.setOnMenuItemClickListener { _ ->
            fragment.lifecycleScope.launch(ioDispatcher) { medicineExport(true) }
            true
        }
        item = menu.findItem(R.id.export_medicine_pdf)
        item.setOnMenuItemClickListener { _ ->
            fragment.lifecycleScope.launch(ioDispatcher) { medicineExport(false) }
            true
        }
    }

    fun setupGenerateTestData() {
        val item = menu.findItem(R.id.generate_test_data)
        val itemWithEvents = menu.findItem(R.id.generate_test_data_and_events)
        if (BuildConfig.DEBUG) {
            itemWithEvents.isVisible = true
            item.isVisible = true
            val menuItemClickListener = MenuItem.OnMenuItemClickListener { menuItem: MenuItem? ->
                idlingResource.setBusy()
                fragment.lifecycleScope.launch(ioDispatcher) {
                    medicineRepository.deleteAll()
                    generateTestDataFactory.create(menuItem === itemWithEvents).generateTestMedicine()
                    requestScheduleNextNotification(context)
                    idlingResource.setIdle()
                }
                true
            }
            item.setOnMenuItemClickListener(menuItemClickListener)
            itemWithEvents.setOnMenuItemClickListener(menuItemClickListener)
        } else {
            item.isVisible = false
            itemWithEvents.isVisible = false
        }
    }

    private fun setupShowAppIntro() {
        val item = menu.findItem(R.id.show_app_intro)
        if (BuildConfig.DEBUG) {
            item.isVisible = true
            item.setOnMenuItemClickListener { _ ->
                context.startActivity(Intent(context, MedTimerAppIntro::class.java))
                true
            }
        } else {
            item.isVisible = false
        }
    }

    private fun handleTagFilter() {
        if (!hideFilter) {
            fragment.lifecycleScope.launch(ioDispatcher) {
                if (medicineRepository.hasTags()) {
                    try {
                        withContext(mainDispatcher) {
                            setupTagFilter()
                        }
                    } catch (_: IllegalStateException) {
                        // Intentionally empty, do nothing
                    }
                } else {
                    medicineViewModel.validTagIds.value = setOf()
                }
            }
        }
    }

    private suspend fun eventExport(isCSV: Boolean) {
        if (medicineViewModel.tagFilterActive()) {
            fragment.lifecycleScope.launch(mainDispatcher) {
                Toast.makeText(context, R.string.tag_filter_active, Toast.LENGTH_LONG).show()
            }
        }
        val reminderEvents: List<ReminderEvent> =
            medicineViewModel.filterEvents(medicineRepository.allReminderEventsWithoutDeletedAndAcknowledged)
        val exporter = if (isCSV) csvEventExportFactory.create(reminderEvents, fragment.getParentFragmentManager()) else pdfEventExportFactory.create(
            reminderEvents,
            fragment.getParentFragmentManager()
        )
        export(exporter)
    }

    private suspend fun medicineExport(isCSV: Boolean) {
        if (medicineViewModel.tagFilterActive()) {
            fragment.lifecycleScope.launch(mainDispatcher) {
                Toast.makeText(context, R.string.tag_filter_active, Toast.LENGTH_LONG).show()
            }
        }
        val medicines: List<FullMedicine> = medicineViewModel.filterMedicines(medicineRepository.medicines)
        val exporter = if (isCSV) csvMedicineExportFactory.create(medicines, fragment.getParentFragmentManager()) else pdfMedicineExportFactory.create(
            medicines,
            fragment.getParentFragmentManager()
        )
        export(exporter)
    }

    private fun setupTagFilter() {
        if (fragment.view == null) {
            return
        }

        val item = menu.findItem(R.id.tag_filter)
        item.isVisible = true
        item.setOnMenuItemClickListener { _ ->
            val tagDataFromPreferences = TagDataFromPreferences(fragment)
            val dialog: DialogFragment = tagsFragmentFactory.create(tagDataFromPreferences)
            dialog.show(fragment.getParentFragmentManager(), "tags")
            true
        }
        fragment.viewLifecycleOwner.lifecycleScope.launch {
            medicineViewModel.validTagIds.collect { validTagIds ->
                if (validTagIds.isNullOrEmpty()) {
                    item.setIcon(R.drawable.tag)
                } else {
                    item.setIcon(R.drawable.tag_fill)
                }
            }
        }
    }

    private suspend fun export(export: Export) {
        val csvFile = File(context.cacheDir, getExportFilename(export))
        try {
            export.export(csvFile)
            FileHelper.shareFile(context, csvFile)
        } catch (_: ExporterException) {
            Toast.makeText(context, R.string.export_failed, Toast.LENGTH_LONG).show()
        }
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        return false
    }

    override fun onDestroy() {
        idlingResource.destroy()
    }

    companion object {
        @SuppressLint("RestrictedApi")
        fun enableOptionalIcons(menu: Menu) {
            if (menu is MenuBuilder) {
                try {
                    val m = menu.javaClass.getDeclaredMethod(
                        "setOptionalIconsVisible", Boolean::class.java
                    )
                    m.invoke(menu, true)
                } catch (_: NoSuchMethodException) {
                    // Intentionally empty
                } catch (_: IllegalAccessException) {
                    // Intentionally empty
                } catch (_: InvocationTargetException) {
                    // Intentionally empty
                }
            }
        }
    }
}
